File: corelibrary/src/main/java/pers/cxd/corelibrary/util/reflection/MethodKey.java
package pers.cxd.corelibrary.util.reflection;

import java.util.Arrays;
import java.util.Objects;

public class MethodKey {

    private static final int MAX_POOL_SIZE = 10;
    private static MethodKey sPool;
    private static int sPoolSize;
    Class<?> clazz;
    String methodName;
    Class<?>[] parameterTypes;
    private boolean inUse;
    private MethodKey next;

    private MethodKey(Class<?> clazz, String methodName, Class<?>[] parameterTypes) {
        this.clazz = clazz;
        this.methodName = methodName;
        this.parameterTypes = parameterTypes;
    }

    public static MethodKey obtain(Class<?> clazz, String methodName, Class<?>[] parameterTypes) {
        synchronized (MethodKey.class) {
            if (sPool != null) {
                MethodKey mk = sPool;
                sPool = mk.next;
                mk.next = null;
                sPoolSize--;
                mk.clazz = clazz;
                mk.methodName = methodName;
                mk.parameterTypes = parameterTypes;
                return mk;
            }
        }
        return new MethodKey(clazz, methodName, parameterTypes);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        MethodKey methodKey = (MethodKey) o;
        return Objects.equals(clazz, methodKey.clazz) &&
                Objects.equals(methodName, methodKey.methodName) &&
                Arrays.equals(parameterTypes, methodKey.parameterTypes);
    }

    @Override
    public int hashCode() {
        int result = Objects.hash(clazz, methodName);
        result = 31 * result + Arrays.hashCode(parameterTypes);
        return result;
    }

    public void markInUse() {
        inUse = true;
    }

    public void recycle() {
        if (inUse) {
            throw new IllegalStateException(this + " is in use, can't recycle");
        }
        clazz = null;
        methodName = null;
        parameterTypes = null;
        synchronized (MethodKey.class) {
            if (sPoolSize < MAX_POOL_SIZE) {
                next = sPool;
                sPool = this;
                sPoolSize++;
            }
        }
    }

}

File: corelibrary/src/main/java/pers/cxd/corelibrary/util/reflection/MethodKey.kt
package pers.cxd.corelibrary.util.reflection

import java.util.*

class MethodKey private constructor(
    var clazz: Class<*>?,
    var methodName: String?,
    var parameterTypes: Array<out Class<*>?>
) {
    private var inUse = false
    private var next: MethodKey? = null
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val methodKey = other as MethodKey
        return clazz == methodKey.clazz &&
                methodName == methodKey.methodName &&
                parameterTypes.contentEquals(methodKey.parameterTypes)
    }

    override fun hashCode(): Int {
        var result = Objects.hash(clazz, methodName)
        result = 31 * result + parameterTypes.contentHashCode()
        return result
    }

    fun markInUse() {
        inUse = true
    }

    fun recycle() {
        check(!inUse) { "$this is in use, can't recycle" }
        clazz = null
        methodName = null
        parameterTypes = emptyArray()
        synchronized(MethodKey::class.java) {
            if (sPoolSize < MAX_POOL_SIZE) {
                next = sPool
                sPool = this
                sPoolSize++
            }
        }
    }

    companion object {
        private const val MAX_POOL_SIZE = 10
        private var sPool: MethodKey? = null
        private var sPoolSize = 0
        fun obtain(
            clazz: Class<*>,
            methodName: String,
            vararg parameterTypes: Class<*>?
        ): MethodKey {
            synchronized(MethodKey::class.java) {
                if (sPool != null) {
                    val mk = sPool
                    sPool = mk!!.next
                    mk.next = null
                    sPoolSize--
                    mk.clazz = clazz
                    mk.methodName = methodName
                    mk.parameterTypes = parameterTypes
                    return mk
                }
            }
            return MethodKey(clazz, methodName, parameterTypes)
        }
    }
}